### ISSUE TITLE
Regression: kube-up failure rates in e2e on GCE increased 6x since yesterday

### ISSUE THREAD

**Issue by quinton-hoole** [kubernetes/kubernetes#23461]
There is a noticeable increase in the number of clusters that fail to come up correctly in our GCE e2e CI tests:

http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gce-ubernetes-lite/

Today there have been 8 such failures in 16 hours (1 per 2 hours).

Yesterday there were 2 failures in 24 hours (1 per 12 hours).

Looking at the logs, it's not clear exactly what's failing. Further investigation required.

```
14:41:58 ... Starting cluster in us-central1-f using provider gce
14:41:58 ... calling verify-prereqs
14:41:58 
14:41:58 All components are up to date.
14:41:59 
14:41:59 All components are up to date.
14:42:00 
14:42:00 All components are up to date.
14:42:00 ... calling kube-up
14:42:00 Project: k8s-jkns-e2e-gce-ubelite
14:42:00 Zone: us-central1-f
14:42:02 +++ Staging server tars to Google Storage: gs://kubernetes-staging-1223c81848/jenkins-e2e-devel
14:42:08 +++ kubernetes-server-linux-amd64.tar.gz uploaded (sha1 = f7e7d74f9f8bfac4630527a3f8b7024db72134aa)
14:42:09 2016/03/24 14:42:09 e2e.go:200: Error running up: exit status 1
14:42:09 2016/03/24 14:42:09 e2e.go:196: Step 'up' finished in 4m51.115019648s
14:42:09 
2016/03/24 14:42:09 e2e.go:110: Error starting e2e cluster. Aborting.
```

**Comment by quinton-hoole** [comment]
I see a similar trend in normal (non-ubernetes) cluster turnups, so the problem does not seem to be ubernetes-specific.  kube-up changes? e.g:

http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gce/

10/24 - 7 failures in 16 hours ~ MTTF 137 min
10/23 - 6 failures in 24 hours ~ MTTF 240 min
10/22 - 1 failures in 24 hours ~ MTTF 1440 min

**Comment by quinton-hoole** [comment]
@mml build cop FYI.  Looks like we might need to roll something back.

**Comment by mml** [comment]
So I'm looking for something merged to cluster (probably) any time during 3/23.

**Comment by mml** [comment]
```
02:14:32 ... calling kube-up
02:14:32 Project: k8s-jkns-e2e-gce-ubelite
02:14:32 Zone: us-central1-f
02:14:34 Creating gs://kubernetes-staging-1223c81848
02:14:34 Creating gs://kubernetes-staging-1223c81848/...
02:14:35 ServiceException: 409 Bucket kubernetes-staging-1223c81848 already exists.
02:14:35 2016/03/24 02:14:35 e2e.go:200: Error running up: exit status 1
02:14:35 2016/03/24 02:14:35 e2e.go:196: Step 'up' finished in 5m0.383120722s
02:14:35 
2016/03/24 02:14:35 e2e.go:110: Error starting e2e cluster. Aborting.
```

**Comment by mml** [comment]
We enter the code that creates the bucket only if 'gsutil ls' returns nonzero.  That could happen on an error condition as well, or perhaps there is a race with another run.  We swallow stdout and stderr from the prior command.

https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/util.sh#L226

**Comment by quinton-hoole** [comment]
Ping @mml .  Any progress on this?  Failure rates still seem to be very high.  If you're unable to work on this now, perhaps we should re-assign to current buildcop?

**Comment by mml** [comment]
@quinton-hoole was out Mon-Tue, but I'll take a look again today.

**Comment by mml** [comment]
I am having some trouble parsing this.  It looks like we got the cluster up, then immediately called kube-up again and failed a few seconds later.

http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gce-ubernetes-lite/1381/consoleFull#4414052197e9f59be-260c-4db5-b752-a7841063d0d0

```
04:43:33 ... calling kube-up
[...]
04:46:48 Cluster validation succeeded
04:46:48 Done, listing cluster services:
[...]
04:46:49 
04:46:49 ... Starting cluster in us-central1-b using provider gce
04:46:49 ... calling verify-prereqs
04:46:49 
04:46:49 All components are up to date.
04:46:49 
04:46:49 All components are up to date.
04:46:50 
04:46:50 All components are up to date.
04:46:50 ... calling kube-up
04:46:50 Project: k8s-jkns-e2e-gce-ubelite
04:46:50 Zone: us-central1-b
04:46:52 +++ Staging server tars to Google Storage: gs://kubernetes-staging-1223c81848/jenkins-e2e-devel
04:46:56 2016/03/28 04:46:56 e2e.go:200: Error running up: exit status 1
04:46:56 2016/03/28 04:46:56 e2e.go:196: Step 'up' finished in 3m24.13465366s
04:46:56 
2016/03/28 04:46:56 e2e.go:110: Error starting e2e cluster. Aborting.
04:46:56 exit status 1
```

**Comment by mml** [comment]
Upon re-reading the log, I see that the first cluster came up and we'd just started on the second.  No sign of the error, though.  We need #23504 to see clearly.

**Comment by quinton-hoole** [comment]
@mml  I'd suggest focussing on kubernetes-e2e-gce, which has a simpler build sequence, and equivalent failure rate:

http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gce/

The ubernetes-lite variant calls kube-up multiple times, which might make the logs more confusing for debugging purposes.

**Comment by mml** [comment]
In http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gce/14244/ the cluster came up but it then refused the connection.  kube-apiserver.log doesn't indicate any reason why it would, but in this case there's a brittle non-retry step that we failed, aborting everything.

```
08:44:22 Waiting for 4 ready nodes. 1 ready nodes, 4 registered. Retrying.
08:44:38 Found 4 node(s).
08:44:38 The connection to the server 104.197.183.165 was refused - did you specify the right host or port?
08:44:38 2016/04/04 08:44:38 e2e.go:200: Error running up: exit status 1
08:44:38 2016/04/04 08:44:38 e2e.go:196: Step 'up' finished in 3m34.349463262s
```

Last line of kube-apiserver.log:

```
 I0404 15:44:47.919786       9 handlers.go:159] PUT /api/v1/nodes/jenkins-e2e-minion-34xt/status: (8.618104ms) 200 [[kubelet/v1.3.0 (linux/amd64) kubernetes/5470488] 10.240.0.3:60273]
```

IN http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gce/14238/, http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gce/14231/, and http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-e2e-gce/14226/ gcloud crashed and that's another place where we won't retry.

```
05:40:53 +++ Staging server tars to Google Storage: gs://kubernetes-staging-b19291185a/jenkins-e2e-devel
05:40:58 +++ kubernetes-server-linux-amd64.tar.gz uploaded (sha1 = df1ed3699374dd08d8e9367749cf8e92c0662c71)
05:41:00 +++ kubernetes-salt.tar.gz uploaded (sha1 = 649ca9fb122510358943b07cfdbd669fa36f7683)
05:41:01 Starting master and configuring firewalls
05:41:01 ERROR: gcloud crashed (error): [Errno 111] Connection refused
05:41:01 
05:41:01 If you would like to report this issue, please run the following command:
05:41:01   gcloud feedback
05:41:01 2016/04/04 05:41:01 e2e.go:200: Error running up: exit status 1
05:41:01 2016/04/04 05:41:01 e2e.go:196: Step 'up' finished in 11.534286952s
```

We maybe should have better retry logic around gcloud invocations, and we should definitely file a bug against gcloud.

**Comment by mml** [comment]
[Another one](https://pantheon.corp.google.com/storage/browser/kubernetes-jenkins/logs/kubernetes-e2e-gce/14348) failing with the apiserver just disappearing.  I see no evidence on the server logs that anything went wrong.

```
11:44:54 Validate output:
11:44:55 NAME                 STATUS    MESSAGE              ERROR
11:44:55 scheduler            Healthy   ok                   
11:44:55 controller-manager   Healthy   ok                   
11:44:55 etcd-0               Healthy   {"health": "true"}   
11:44:55 etcd-1               Healthy   {"health": "true"}   
11:44:55 Cluster validation succeeded
11:44:55 Done, listing cluster services:
11:44:55 
11:44:55 Unable to connect to the server: EOF
11:44:55 2016/04/06 11:44:55 e2e.go:200: Error running up: exit status 1
11:44:55 2016/04/06 11:44:55 e2e.go:196: Step 'up' finished in 3m59.820652119s
```

**Comment by mml** [comment]
So #23842 is my approach to the gcloud problem.  For the mysteriously disappearing apiserver problem, I am going to run a loop overnight that just brings up and tears down an e2e cluster over and over again hoping to replicate.  The loop will stop if 'up' fails, and hopefully I'll better be able to debug the half finished cluster since nothing will tear it down.

**Comment by mml** [comment]
OK, here we go.  I was able to reproduce this and the cluster is still intact.

```
Found 5 node(s).
NAME              STATUS                     AGE
e2e-master        Ready,SchedulingDisabled   46s
e2e-minion-6jvi   Ready                      45s
e2e-minion-erw7   Ready                      47s
e2e-minion-iw1t   Ready                      44s
e2e-minion-l57z   Ready                      49s
Unable to connect to the server: EOF
Validate output:
The connection to the server 104.197.251.20 was refused - did you specify the right host or port?
2016/04/06 17:50:50 e2e.go:200: Error running up: exit status 1
2016/04/06 17:50:50 e2e.go:196: Step 'up' finished in 3m51.590921585s
2016/04/06 17:50:50 e2e.go:110: Error starting e2e cluster. Aborting.
```

(I have set the TZ to match my workstation... 17:50 is the same as 00:50 UTC.)

```console
mml@e2e-master:/var/log$ ps -o stime,start,pid,cmd $(pidof kube-apiserver)
STIME  STARTED   PID CMD
Apr06 17:50:52  5274 /usr/local/bin/kube-apiserver --address=127.0.0.1 --etcd-servers=http://127.0.0.1:4001 --etcd-servers-overrides=/events#http://127.0.0.1:4002 --cloud-provider=gce --admission-con
```

So it does look like the apiserver was restarted at about the correct time.  And... here it is.  Our last log line at 00:50:49, and then a new process

```
I0407 00:50:49.932455       7 handlers.go:159] GET /api/v1/watch/pods?fieldSelector=spec.nodeName%3D%2Cstatus.phase%21%3DFailed%2Cstatus.phase%21%3DSucceeded&resourceVersion=3&timeoutSeconds=481: (49.616611626s) 200 [[kube-scheduler/v1.2.0 (linux/amd64) kubernetes/3a4beb1] 127.0.0.1:38511]
Flag --address has been deprecated, see --insecure-bind-address instead
I0407 00:50:53.094286       6 server.go:192] Will report 10.240.0.2 as public IP address.
I0407 00:50:53.100302       6 gce.go:268] Using existing Token Source &oauth2.reuseTokenSource{new:google.computeSource{account:""}, mu:sync.Mutex{state:0, sema:0x0}, t:(*oauth2.Token)(nil)}
```

No note of why it died, but docker was restarted at about the same time (switching back to UTC).

```
mml@e2e-master:/var/log/supervisor$ ps -o start,pid,cmd $(pidof docker)
 STARTED   PID CMD
00:50:48  5089 /usr/bin/docker daemon -p /var/run/docker.pid --bridge=cbr0 --iptables=false --ip-masq=false --log-level=info
```

and in docker.log...

```
time="2016-04-07T00:50:52.600029705Z" level=info msg="POST /containers/create?name=k8s_kube-apiserver.f8d03ba6_kube-apiserver-e2e-master_kube-system_e00564977a51bc22923bdce1015a0a96_ea5bab6f" 
```

So for some reason docker exited, and on restart it restarted the apiserver, and our script can easily hit this window causing it to bail.  The script is too brittle, but I'd also like to find out what's happening to docker.

**Comment by mml** [comment]
Docker is started by an init script.  After that, supervisord is started, and the docker-checker script begins by killing then restarting any running docker.  The timestamps on the files support that this happened.  This pretty much guarantees one restart close to boot time, and if that happens *after* our 300 second wait for cluster init, we may try to contact the server during the few seconds when it's restarting.

As a kludge to reduce the incidence of this, we can modify the docker-checker script to not begin with a kill.  Ultimately either kubectl or our shell scripts (please no!) need to be more relaxed and willing to retry.  I'll file a separate bug for that.

**Comment by quinton-hoole** [comment]
Awesome debugging @mml.  Thanks!

**Comment by quinton-hoole** [comment]
@mml I wonder whether we upgraded to a new docker version at the time when this failure rate increased so markedly?

**Comment by mml** [comment]
@quinton-hoole I was trying to figure that out, too.  There has been some churn in the docker-checker.sh scripts with extra sleeps added at various times since February trying to make sure docker is really dead before we remove its networking checkpoint.  Those have had the effect of inserting more delay between the stopping and starting of docker itself.

https://github.com/kubernetes/kubernetes/blame/master/cluster/saltbase/salt/supervisor/docker-checker.sh

I find it hard to detangle the true time of change because everything is marked by when the commit was authored rather than when it was merged.

**Comment by mml** [comment]
I think something else reduced the errors we were seeing from gcloud.  Haven't seen those since.

And we just [finally] merged #23992 which should make failure to start much rarer.

**Comment by mml** [comment]
I am going to keep my eye on this.  We just had another failure to start in e2e-gce-slow.  In this case, we spent the whole time trying to download the binaries from GCS and by the time 5m was up we still hadn't finished.  This is a new failure mode that is probably rare.

But considering it'll now take an hour for the slow test suite to go green again, this really sucks.

 - https://console.cloud.google.com/storage/browser/kubernetes-jenkins/logs/kubernetes-e2e-gce-slow/4164
 - [startupscript.log](https://00e9e64baccb5bc24fd8b45006cfdd9e8b12e3ce8752817be6-apidata.googleusercontent.com/download/storage/v1_internal/b/kubernetes-jenkins/o/logs%2Fkubernetes-e2e-gce-slow%2F4164%2Fartifacts%2Fjenkins-e2e-master%2Fstartupscript.log?qk=AD5uMEuCQrLdDBngAEXQxBaGt7QKowvj002YfqxALaIRyASOaTuEDZeXi2V9lpGw3s9JbnWLAxMDNeNF3Pwf1zHaUji_SMtO04ghinp-jybPa37nepTOVUoH4DxW0MXfnFCjTp27wJpgRLOEhloq-Q3X2dpYf7j9CdV9_0kdK6OOGBqqgFoAgXNpHQAlywauOWOJ4WdsyzginHiQRL1NVQQarIy2tv5GQ3mRsqgr0kf77hEBZ-jka2Vay7Suvb6LVijm7X5czjpCY5ntnT2VjeN_KuQzJ13AaigMwKTyjh6QPdgIe5Z0kBlDVAEz4qDJVujvHnfmemtOI9pMsZQnOkrF-VK23pHLmDbleV2L-vbT4tfln358HtBC6wIRf1KpUBUANL6leCa9kYKTGyJHXq_NPUOX24OpIytpSzBzDpZKlPkjFpZc4dFWTEgeJI3ZBxxV8akL0ePhglSJXUEalPg2MmLDgXQl53ZeaGfAaOUSHWta_yMubw3tcNujjr93XZiwpdMcHFju-jkbkoGB9_Cr7vgFJshIkb5VWlYSQuXol1regu-AhenWzy68VSwSqfSgAL5J4esyBFDKhHzxxrqjGhSGGXfhY2MdLyvSs1I8nqMfneLMGH9HgKuyLmRBZnhfPp2tJfsx76HQA14PVpai3gvmkF1Uy0PrfXUGOpjib_SOoB6jYFJIxOa4831C-4ZtDM9WD-IOBPEX67XhR0HJkQ6v7qvXxW4z1QdQGa4O5BbPuX9CUGc222OkLx91viMYseVtHh-Wgvtvbm3a_HHvOuPmSSkOyfX6REJrZiwr6ihlozsiswcu8N-Ov41Uv7ayyDw4uYOe5cf1OO4jt99CLCQX9HBxPGV_YVBJsOxxkCgLTdt6BRMrJGstLME4hALQ-tt0H47CJ7w3yl6U6SAySw5O5Zqss6nYVQPb6K9rIYJum5SlMEU)

**Comment by quinton-hoole** [comment]
@mml looks awesome so far.  We've had a 6 hour stretch of green on http://kubekins.dls.corp.google.com/job/kubernetes-e2e-gce-ubernetes-lite/ since you merged #23992.  We haven't had a 6 hour stretch of green on that job in 3 weeks, since I opened this issue :-)

Looks good so far!
